src/Microsoft.DotNet.Interactive.CSharpProject/RoslynWorkspaceUtilities/RoslynWorkspaceUtilities.cs

```diff
@@ -0,0 +1,332 @@
+        WorkspaceProjectReferences[projectId.Id] = analyzerResult.ProjectReferences.ToArray();
+
+        ProjectInfo projectInfo = GetProjectInfo(analyzerResult, workspace, projectId);
+        if (projectInfo is null)
+        {
+            // Error
+            return null;
+        }
+        Solution solution = workspace.CurrentSolution.AddProject(projectInfo);
+
+        if (!workspace.TryApplyChanges(solution))
+        {
+            throw new InvalidOperationException("Could not apply workspace solution changes");
+        }
+
+        return workspace.CurrentSolution.GetProject(projectId);
+    }
+
+    internal static AdhocWorkspace CreateWorkspace()
+    {
+        AdhocWorkspace workspace = new AdhocWorkspace();
+        workspace.WorkspaceChanged += WorkspaceChangedHandler;
+        workspace.WorkspaceFailed += WorkspaceFailedHandler;
+        return workspace;
+    }
+
+    private static void WorkspaceFailedHandler(object sender, WorkspaceDiagnosticEventArgs e)
+    {
+        // Log error
+    }
+
+    private static void WorkspaceChangedHandler(object sender, WorkspaceChangeEventArgs e)
+    {
+        // Log
+    }
+
+    private static ProjectInfo GetProjectInfo(BuildProjectData buildProjectData, CodeAnalysis.Workspace workspace, ProjectId projectId)
+    {
+        string projectName = Path.GetFileNameWithoutExtension(buildProjectData.ProjectFilePath);
+        if (!TryGetSupportedLanguageName(buildProjectData.ProjectFilePath, out string languageName))
+        {
+            return null;
+        }
+        return ProjectInfo.Create(
+            projectId,
+            VersionStamp.Create(),
+            projectName,
+            projectName,
+            languageName,
+            filePath: buildProjectData.ProjectFilePath,
+            outputFilePath: buildProjectData.TargetPath,
+            documents: GetDocuments(buildProjectData, projectId),
+            projectReferences: GetExistingProjectReferences(buildProjectData, workspace),
+            metadataReferences: GetMetadataReferences(buildProjectData),
+            analyzerReferences: GetAnalyzerReferences(buildProjectData, workspace),
+            parseOptions: CreateParseOptions(buildProjectData, languageName),
+            compilationOptions: CreateCompilationOptions(buildProjectData, languageName));
+    }
+
+    private static bool TryGetSupportedLanguageName(string projectPath, out string languageName)
+    {
+        switch (Path.GetExtension(projectPath))
+        {
+            case ".csproj":
+                languageName = LanguageNames.CSharp;
+                return true;
+            default:
+                languageName = null;
+                return false;
+        }
+    }
+
+    private static IReadOnlyList<DocumentInfo> GetDocuments(BuildProjectData buildProjectData, ProjectId projectId) =>
+        buildProjectData
+            .SourceFiles?
```

Comment:
This method should not be allowed to return null. If there are no source files, we should return an empty list.

Reply:
fixed